L1Trigger/TrackFindingTMTT/src/KFParamsComb.cc

```diff
@@ -0,0 +1,465 @@
+///=== This is the Kalman Combinatorial Filter for 4 helix parameters track fit algorithm.
+
+#include "L1Trigger/TrackFindingTMTT/interface/KFParamsComb.h"
+#include "L1Trigger/TrackFindingTMTT/interface/KalmanState.h"
+#include "L1Trigger/TrackFindingTMTT/interface/StubCluster.h"
+#include "DataFormats/Math/interface/deltaPhi.h"
+//#define CKF_DEBUG
+
+namespace tmtt {
+
+  /*
+// Scattering constants - HISTORIC NOT USED.
+
+constexpr unsigned nlayer_eta[25] = 
+{ 6, 6, 6, 6,
+6, 6, 6, 6, 6, 6, 6, 7, 7, 7,
+7, 7, 7, 7, 6, 6, 6, 6, 6, 6};
+
+constexpr double matx_outer[25] = {
+0.16, 0.17, 0.18, 0.19, 0.20, 
+0.21, 0.26, 0.22, 0.26, 0.38,
+0.41, 0.40, 0.44, 0.50, 0.54,
+0.60, 0.44, 0.48, 0.60, 0.68,
+0.50, 0.48, 0.64, 0.39, 0.20
+};
+
+constexpr double matx_inner[25] = {
+0.14, 0.1, 0.1, 0.1, 0.1, 
+0.1, 0.1, 0.1, 0.1, 0.1, 
+0.12, 0.1, 0.1, 0.1, 0.15,
+0.20, 0.25, 0.25, 0.3, 0.3,
+0.35, 0.40, 0.40, 0.6, 0.6
+};
+*/
+
+  KFParamsComb::KFParamsComb(const Settings* settings, const uint nPar, const string& fitterName)
+      : L1KalmanComb(settings, nPar, fitterName) {
+    hdxmin[INV2R] = -1.1e-4;
+    hdxmax[INV2R] = +1.1e-4;
+    hdxmin[PHI0] = -6.e-3;
+    hdxmax[PHI0] = +6.e-3;
+    hdxmin[Z0] = -4.1;
+    hdxmax[Z0] = +4.1;
+    hdxmin[T] = -6.;
+    hdxmax[T] = +6.;
+    if (nPar_ == 5) {
+      hdxmin[D0] = -1.001;
+      hdxmax[D0] = +1.001;
+    }
+
+    hxmin[INV2R] = -0.3 * 0.0057;
+    hxmax[INV2R] = +0.3 * 0.0057;
+    hxmin[PHI0] = -0.3;
+    hxmax[PHI0] = +0.3;
+    hxmin[Z0] = -120;
+    hxmax[Z0] = +120;
+    hxmin[T] = -6.;
+    hxmax[T] = +6.;
+    if (nPar_ == 5) {
+      hxmin[D0] = -3.5;
+      hxmax[D0] = +3.5;
+    }
+
+    hddMeasmin[PHI0] = -1.e1;
+    hddMeasmax[PHI0] = +1.e1;
+
+    hresmin[PHI0] = -0.5;
+    hresmax[PHI0] = +0.5;
+
+    hresmin[PHI0] = -10.;
+    hresmax[PHI0] = +10.;
+
+    hxaxtmin[INV2R] = -1.e-3;
+    hxaxtmax[INV2R] = +1.e-3;
+    hxaxtmin[PHI0] = -1.e-1;
+    hxaxtmax[PHI0] = +1.e-1;
+    hxaxtmin[Z0] = -10.;
+    hxaxtmax[Z0] = +10.;
+    hxaxtmin[T] = -1.e-0;
+    hxaxtmax[T] = +1.e-0;
+    if (nPar_ == 5) {
+      hxaxtmin[D0] = -1.001;
+      hxaxtmax[D0] = +1.001;
+    }
+  }
+
+  std::map<std::string, double> KFParamsComb::getTrackParams(const KalmanState* state) const {
+    std::vector<double> x = state->xa();
+    std::map<std::string, double> y;
+    y["qOverPt"] = 2. * x.at(INV2R) / getSettings()->invPtToInvR();
+    y["phi0"] = reco::deltaPhi(x.at(PHI0) + sectorPhi(), 0.);
+    y["z0"] = x.at(Z0);
+    y["t"] = x.at(T);
+    if (nPar_ == 5) {
+      y["d0"] = x.at(D0);
+    }
+    return y;
+  }
+
+  /* If using 5 param helix fit, get track params with beam-spot constraint & track fit chi2 from applying it. */
+  /* (N.B. chi2rz unchanged by constraint) */
+
+  std::map<std::string, double> KFParamsComb::getTrackParams_BeamConstr(const KalmanState* state,
+                                                                        double& chi2rphi) const {
+    if (nPar_ == 5) {
```

Comment:
The fitter can either do 4 or 5 parameter helix fits (depending on if constraint d0 = 0 is applied). Lines such as (nPar = 5) are checking if a 5 parameter fit is being done. I think replacing 5 by a named constant would make the code more obscure.

Reply:
I agree that `5` makes more sense than `FREE_PARAMETERS_WITHOUT_D0_CONSTRAINT` ...

Do you think it would help readability adding a one-line comment where this is used, something like
```suggestion
    if (nPar_ == 5) {  // no d0 constraint
```
?

Reply:
It's not just to help comprehension, but to ensure consistency.

Here's an alternate suggestion in this case: define `bool is5ParFit = (nPar_ == 5)` once (maybe even as a class member), and then use it everywhere. (Same for 4-param fit.)

Reply:
I dislike the proposed solutions. They add to the complexity of the code, without improving its comprehensibility. And in an expression such as "if (nPar == 5)", the variable name at the left-hand side of the equality sign makes clear what the meaning of the number 5 is.
As a compromise, I've added the comment suggested above. And have also renamed "nPar" to "nHelixPar" to make its meaning clearer.